presto-main/src/main/java/com/facebook/presto/operator/LongArrayBlockEncodingBuffers.java

```diff
@@ -0,0 +1,188 @@
+        if (batchSize == 0) {
+            return;
+        }
+
+        verify(decodedBlock != null);
+
+        int[] positions = getPositions();
+        verify(positionsOffset + batchSize - 1 < positions.length && positions[positionsOffset] >= 0 &&
+                positions[positionsOffset + batchSize - 1] < decodedBlock.getPositionCount());
+
+        appendValuesToBuffer();
+        appendNulls();
+        bufferedPositionCount += batchSize;
+    }
+
+    @Override
+    public void serializeTo(SliceOutput sliceOutput)
+    {
+        // TODO: getSerializedSizeInBytes() was calculated in flush and doesnt need to recalculated
+        verify(getSerializedSizeInBytes() <= sliceOutput.writableBytes());
+
+        writeLengthPrefixedString(sliceOutput, NAME);
+        sliceOutput.writeInt(bufferedPositionCount);
+        serializeNullsTo(sliceOutput);
+        sliceOutput.appendBytes(valuesBuffer, 0, valuesBufferIndex);
+    }
+
+    @Override
+    public long getSizeInBytes()
+    {
+        return getPositionsSizeInBytes() + // positions and mappedPositions
+                valuesBufferIndex +
+                getNullsBufferSizeInBytes();
+    }
+
+    @Override
+    public long getRetainedSizeInBytes()
+    {
+        return INSTANCE_SIZE +
+                getPostionsRetainedSizeInBytes() +
+                sizeOf(valuesBuffer) +
+                getNullsBufferRetainedSizeInBytes();
+    }
+
+    @Override
+    public long getSerializedSizeInBytes()
+    {
+        return NAME.length() + SIZE_OF_INT +    // NAME
+                SIZE_OF_INT +                   // positionCount
+                getNullsBufferSerializedSizeInBytes() +   // nulls buffer
+                valuesBufferIndex;              // values buffer
+    }
+
+    private void appendValuesToBuffer()
+    {
+        verify(decodedBlock != null);
+
+        ensureValueBufferSize();
+
+        if (decodedBlock.mayHaveNull()) {
+            appendLongsWithNullsToBuffer();
+        }
+        else {
+            appendLongsToBuffer();
+        }
+    }
+
+    private void appendLongsToBuffer()
+    {
+        int[] positions = getPositions();
+
+        for (int i = positionsOffset; i < positionsOffset + batchSize; i++) {
+            long value = decodedBlock.getLong(positions[i]);
+            ByteArrayUtils.writeLong(valuesBuffer, valuesBufferIndex, value);
+            valuesBufferIndex += ARRAY_LONG_INDEX_SCALE;
+        }
+    }
+
+    private void appendLongsWithNullsToBuffer()
+    {
+        int[] positions = getPositions();
+
+        for (int i = positionsOffset; i < positionsOffset + batchSize; i++) {
+            int position = positions[i];
+            long value = decodedBlock.getLong(position);
+            ByteArrayUtils.writeLong(valuesBuffer, valuesBufferIndex, value);
+
+            if (!decodedBlock.isNull(position)) {
+                valuesBufferIndex += ARRAY_LONG_INDEX_SCALE;
+            }
+        }
+    }
+
+    private void ensureValueBufferSize()
```

Comment:
- ensureValueBufferSize -> ensureValuesBufferCapacity for consistency
- requiredSize -> capacity
- swap the sides of `<`

```
    private void ensureValuesBufferCapacity()
    {
        int capacity = valuesBufferIndex + batchSize * ARRAY_LONG_INDEX_SCALE;

        if (valuesBuffer.length < capacity) {
            valuesBuffer = Arrays.copyOf(valuesBuffer, max(valuesBuffer.length * 2, capacity));
        }
    }
```